### ISSUE TITLE
FLOPs compute

### ISSUE THREAD

**Issue by leosk-knight** [Tencent/PocketFlow#87]

**Comment by psyyz10** [comment]
@leosk-knight  It's the ordinary formula.

**Comment by leosk-knight** [comment]
The same is true for MatMul？

**Comment by jiaxiang-wu** [comment]
@leosk-knight 
Previously, we divide the FLOPs by 2 to match the results we obtained from a Caffe-based script for computational complexity. Later, we discover that TensorFlow uses FLOPs to measure the complexity, while that Caffe-based script uses MACs. For most operations in CNN, 1MACs approximately equals to 2FLOPs. Therefore, the division by 2 can be removed.

Below is a detailed script to demonstrate that for Conv2D and MatMul, TensorFlow counts both multiplication and addition as one FLOPs, so the overall FLOPs is approximately two times of MACs.

Scripts:
``` Python
import tensorflow as tf

# Conv2d
batch_size = 16
img_hei = 32
img_wid = 32
img_chn = 8
nb_chns_out = 64
kernel_size = 3
with tf.Graph().as_default() as g:
  x = tf.zeros([batch_size, img_hei, img_wid, img_chn])
  y = tf.layers.conv2d(x, nb_chns_out, kernel_size, strides=1, padding='same')
  run_meta = tf.RunMetadata()
  opts = tf.profiler.ProfileOptionBuilder.float_operation()
  flops = tf.profiler.profile(g, run_meta=run_meta, cmd='op', options=opts)
  print('TensorFlow says: %d' % flops.total_float_ops)
  print('our estimation: %d' % (2 * batch_size * img_hei * img_wid * nb_chns_out * kernel_size ** 2 * img_chn))

# MatMul
batch_size = 16
nb_dims_in = 128
nb_dims_out = 64
with tf.Graph().as_default() as g:
  x = tf.zeros([batch_size, nb_dims_in])
  y = tf.layers.dense(x, nb_chns_out)
  run_meta = tf.RunMetadata()
  opts = tf.profiler.ProfileOptionBuilder.float_operation()
  flops = tf.profiler.profile(g, run_meta=run_meta, cmd='op', options=opts)
  print('TensorFlow says: %d' % flops.total_float_ops)
  print('our estimation: %d' % (2 * batch_size * nb_dims_in * nb_dims_out))
```

Selected output (with a few comments):
``` plain
# tf.layers.conv2d
Profile:
node name | # float_ops
Conv2D                   150.99m float_ops (100.00%, 99.30%)
BiasAdd                  1.05m float_ops (0.70%, 0.69%)
Add                      4.61k float_ops (0.01%, 0.00%)
Mul                      4.61k float_ops (0.00%, 0.00%)
Sub                          1 float_ops (0.00%, 0.00%)

TensorFlow says: 152052737
our estimation: 150994944  # this matches <Conv2D>

# tf.layers.dense
Profile:
node name | # float_ops
MatMul                   262.14k float_ops (100.00%, 93.77%)
Add                      8.19k float_ops (6.23%, 2.93%)
Mul                      8.19k float_ops (3.30%, 2.93%)
BiasAdd                  1.02k float_ops (0.37%, 0.37%)
Sub                          1 float_ops (0.00%, 0.00%)

======================End of Report==========================
TensorFlow says: 279553
our estimation: 262144  # this matches <MatMul>
```

**Comment by jiaxiang-wu** [comment]
**TODO**: remove the division by 2 in the FLOPs computation.

**Comment by jiaxiang-wu** [comment]
Resolved in PR #132.
